### PR TITLE
gh-108469: Update ast.unparse for unescaped quote support from PEP701 [3.12]

### PR DESCRIPTION
**Pull request by tonybaloney**
* Issue: gh-108469

### CODE REVIEW

**Comment by sunmy2019** [comment]
Sorry that I have taken over.

File: Lib/test/test_unparse.py [hunk 1]
```diff
@@ -502,11 +513,11 @@ def test_class_bases_and_keywords(self):
         self.check_src_roundtrip("class X(*args, **kwargs):\n    pass")
 
     def test_fstrings(self):
-        self.check_src_roundtrip('''f\'\'\'-{f"""*{f"+{f'.{x}.'}+"}*"""}-\'\'\'''')
-        self.check_src_roundtrip('''f"\\u2028{'x'}"''')
+        self.check_src_roundtrip("f'-{f'*{f'+{f'.{x}.'}+'}*'}-'")
+        self.check_src_roundtrip("""f'\\u2028{'x'}'""")
```
Comment (sunmy2019) [inline Lib/test/test_unparse.py]:
```suggestion
        self.check_src_roundtrip("f'\\u2028{'x'}'")
```

**Comment by tonybaloney** [comment]
> Sorry that I have taken over.

You're welcome. I spent hours debugging these private methods and can't figure out which ones need removing or changing

**Comment by sunmy2019** [comment]
Oh, I've found some corner case that previous logic did not handle. Give me some time for a more thorough rewrite.

**Comment by sunmy2019** [comment]
I discovered it was a bug in the original implementation. Opened #108843.

**Comment by sunmy2019** [comment]
Fixed on this branch.

File: Lib/ast.py [hunk 2]
```diff
@@ -1273,16 +1285,12 @@ def _write_fstring_inner(self, node):
 
     def visit_FormattedValue(self, node):
         def unparse_inner(inner):
-            unparser = type(self)(_avoid_backslashes=True)
+            unparser = type(self)(_avoid_backslashes=False)
```
Comment (isidentical) [inline Lib/ast.py]: Do we still need this mode or can we drop `avoid_backslashes` all together?

File: Lib/ast.py [hunk 3]
```diff
@@ -1225,37 +1225,49 @@ def _write_str_avoiding_backslashes(self, string, *, quote_types=_ALL_QUOTES):
...
+                    if c == quote:
+                        if consecutive_quotes == len(quote_type) - 1:
+                            # escape when we see a full `quote_type`
+                            res.append("\\")
+                            consecutive_quotes = 0
+                        else:
+                            consecutive_quotes += 1
+                    else:
+                        consecutive_quotes = 0
+                    res.append(c)
```
Comment (isidentical) [inline Lib/ast.py]: This logic probably belongs somewhere else, something similiar to what we had with the existing f-string unparsing implementation:

https://github.com/python/cpython/blob/21229ce1f0c8fb2e580df62972296d629960f30e/Lib/ast.py#L1182C1-L1218C1

**Comment by isidentical** [comment]
@sunmy2019 if we are interested in quickly shipping a fix for it, I'd avoid changing anything cosmetic and simply dropping the existing `avoid_backslashes` logic all together. It should probably probably work for all the supported cases. We can then simplify the handling of string literals with backslashes in a subsequent PR that only improves the generated code cosmeticly

File: Lib/ast.py (hunk 2 above)
Comment (sunmy2019) [reply Lib/ast.py]: I guess we can drop it.

**Comment by sunmy2019** [comment]
> I'd avoid changing anything cosmetic and simply dropping the existing avoid_backslashes logic all together

That what I tried to do before 1da1d4c162b580e5542ba1c0e06bb5314c9ca3dc.

Then I discovered a bug (#108843) in the original implementation. So I thought I can fix it along the way. 

What do you think?

**Comment by isidentical** [comment]
I think separting the fix for that from the actual PEP 701 support makes sense, and also reduces the possible complexity of changes this PR brings in. Let's first do the cleanup and the support for PEP 701 and we can do the fix later since this seems like the more serious problem.

**Comment by pablogsal** [comment]
Let's try to land this today or tomorrow. I added @Yhg1s to the PR so he is aware of it.

**Comment by sunmy2019** [comment]
> Let's first do the cleanup and the support for PEP 701

I got it reverted.

**Review by pablogsal — APPROVED** [review]

**Comment by pablogsal** [comment]
@sunmy2019 can you make a PR for the PEP 701 changes and ping me, @isidentical and @Yhg1s ?

**Comment by sunmy2019** [comment]
> @sunmy2019 can you make a PR for the fix changes and ping me, @isidentical and @Yhg1s ?

Working on it.

**Pull request merged**